File: perf/zmq-latency-b.cpp
#include <stdexcept>
#include <zmq.hpp>

int main(int argc, char* argv[])
{
  if(argc != 4)
    throw std::runtime_error("Usage: <address> <size> <count>");
  const std::string address = argv[1];
  const int size = ::atoi(argv[2]);
  const int count = ::atoi(argv[3]);

  zmq::context_t context(1);
  zmq::socket_t socket(context, ZMQ_REP);
  socket.bind(address.c_str());

  for(int i = 0; i != count; ++i)
  {
    zmq::message_t message;
    socket.recv(&message, 0);
    if(message.size() != size)
      throw std::runtime_error("message size mismatch");
    socket.send(message, 0);
  }

  return 0;
}

File: perf/zmq-latency-b.py
import optparse
import time
import zmq

parser = optparse.OptionParser()
parser.add_option("--address", default="tcp://127.0.0.1:5555", help="Local address.  Default: %default")
parser.add_option("--size", type="int", default=1, help="Message size.  Default: %default")
parser.add_option("--count", type="int", default=100000, help="Message count.  Default: %default")
parser.add_option("--hwm", type="int", default=100000, help="High water mark.  Default: %default")
(options, arguments) = parser.parse_args()

context = zmq.Context()
socket = context.socket(zmq.REP)
socket.setsockopt(zmq.HWM, options.hwm)
socket.bind(options.address)

for i in range(options.count):
  message = socket.recv()
  if len(message) != options.size:
    raise Exception("message size mismatch")
  socket.send(message)

